### DISCUSSION TITLE
Testing Actions on RelationManagers

### DISCUSSION THREAD

**Discussion by titantwentyone** [filamentphp/filament#3293] [Help]
Looking to test some actions provided on a RelationManager table. My action is:

```
Action::make('admit)
    ->requiresConfirmation()
    ->action(fn($record) => ...);
```
My test, which I adapted from the docs, given that this is in a RelationManager:
```
/** @test */
public function it_will_admit_a_person()
{
    $booking = Booking::factory()->create();

    Livewire::actingAs(User::factory()->create())
        ->test(BookingRelationManager::class, [
            'ownerRecord' => $booking->event
        ])
        ->callTableAction('admit', $booking);

    $booking->refresh();

    $this->assertNotEmpty($booking->admission_info);
}
```
The test fails with:
```
Failed asserting that an event [open-modal] was fired with parameters: {"id":"App\\Filament\\Resources\\EventResource\\RelationManagers\\BookingRelationManager-table-action"}
Failed asserting that false is true.
```
Just out of interest, I removed `requiresConfirmation()` from my action to see if this was the issue but this results in:

```
Failed asserting that 'admit' is not equal to 'admit'.
```
The action works as expected when testing manually. Are there additional steps needed to test actions in RelationManagers?

**Comment by danharrin** [comment]
Have you tried testing actions on List pages?

**Reply by titantwentyone** [reply]
Just now. I had an action on the main `Event` resource table that needed to be tested:
```
Livewire::actingAs(User::factory()->create())
    ->test(ListEvents::class)
    ->callTableAction('register', Event::first())
    ->assertFileDownloaded('event.xlsx');
```
Comes back as green.

**Reply by danharrin** [reply]
Weird, I honestly don't know, can you do some debugging?

**Reply by titantwentyone** [reply]
I've had a brief look at this today. When `->requiresConfirmation()` is used the test fails when trying to access `effects.dispatches` of the components payload. The place where this fails is [here](https://github.com/livewire/livewire/blob/63ad79f9bbad51028e1aaeca774f2f5e22f6578d/src/Testing/Concerns/MakesAssertions.php#L241) in the Livewire source. `data_get($this->payload, 'effects.dispatches')` results in null. From Filament, this stems from [here](https://github.com/filamentphp/filament/blob/c9db1519da7f09a84d50374d343c1969c46106a9/packages/tables/src/Testing/TestsActions.php#L52)

Without `->requiresConfirmation()`, the test fails [here](https://github.com/filamentphp/filament/blob/c9db1519da7f09a84d50374d343c1969c46106a9/packages/tables/src/Testing/TestsActions.php#L45) in Filament when trying to determine the mounted action.

That's as far as I've got in all honesty due to time constraints. Originally I thought this was two separate issues but it feels like the action name and payload that are not being passed along for testing. (Odd since it works when actually using this)

Any thoughts?

Edit: typos/formatting!

**Reply by titantwentyone** [reply]
My bad! My actual code included some logic in `->visible()` which would only allow a user access to this action if a setting was applied elsewhere. I managed to find this out by seeing that the action was in fact disabled in `HasActions` when the action was mounted.

Made me think whether it would be worth adding some further assertions such as `assertIsDisabled(...)`, `assertIsIcon(...)`, `assertIsLabel(...)` etc or some generic assertion to check the logic within anything that utilitizes evaluation of a closure. (This could be expanded for columns and fields perhaps?)

**Reply by danharrin** [reply]
Glad you got it working

I would be resistant to add `assertIsIcon` and `assertIsLabel` as these seem to be very appearance-focused and not useful in determining whether an interface works.

However, `assertIsDisabled()` would be a good (and easy) addition, you just need to copy the existing `hidden` version we have. Can you PR?

**Reply by titantwentyone** [reply]
PR submitted #3432 

With regards to the `assertIsIcon` etc - visual elements - my particular use case is a register of sorts and the icons and colours of the actions change depending on whether people have been admitted/discharged. It's useful for my end clients to see, at a glance, whether people are admitted or not and this is formatted through the use of closures - changing icon, colours and labels. If it's wrong, then it doesn't really help them much. It doesn't directly affect whether the action works or not - that's now fully covered - but it's still something that could be tested to ensure my logic is correct and executing as expected. Filament is so flexible and I'm finding little snippets all the time browsing through the source - would just be great to be able to back up as much as I can when coding with it through tests. Also good if the API changes in the future.

**Reply by danharrin** [reply]
Sure - add those then. Also, please add tests to our test suite (`/tests`) that actually use these new helpers.

Also, you will want to add identical action testing helpers for the Admin panel Page Actions too (there's a separate TestsPages within the admin panel package that holds those testing helpers)

**Reply by titantwentyone** [reply]
Updated PR.

Will work on the other assertions (icons, labels etc) shortly.

**Reply by danharrin** [reply]
I added some different testing helpers last night, looks like there is now a merge conflict in your PR
